### ISSUE TITLE
How to add a customizable offset to the start of an animation (asset)?

### ISSUE THREAD

**Issue by yarres** [inlife/nexrender#122]
Hi,

Here's my use case:
I have one aep project that has the following assets: 
* one video (it will be replaced by another video asset)
* one line (no replacement)
* two texts (only the Source Text will be replaced)
The animation with the texts and the line starts at 0 seconds in the timeline.

I would like to be able to choose the time at which the texts and line assets should be played: for example for one job the texts and line should start at 9 seconds and for another one at 121 seconds.

In other words: the start of the animation of specific assets (all except the video) would be delayed and would not start at 0s but at 9 or 121 seconds in this example.
I have started to experiment but I can't find a solution for now.
How do you think it could be done? Maybe using a .jsx (but then I would have to pass the offset value to the jsx)? Or maybe it's not achievable?

Anyway, thanks for this projects, it seems really powerful :)

**Comment by yarres** [comment]
Hi,

Here's my use case:
I have one aep project that has the following assets: 
* one video (it will be replaced by another video asset)
* one line (no replacement)
* two texts (only the Source Text will be replaced)
The animation with the texts and the line starts at 0 seconds in the timeline.

I would like to be able to choose the time at which the texts and line assets should be played: for example for one job the texts and line should start at 9 seconds and for another one at 121 seconds.

In other words: the start of the animation of specific assets (all except the video) would be delayed and would not start at 0s but at 9 or 121 seconds in this example.
I have started to experiment but I can't find a solution for now.
How do you think it could be done? Maybe using a .jsx? Or maybe it's not achievable?

**Comment by inlife** [comment]
its related to #121

>Maybe using a .jsx (but then I would have to pass the offset value to the jsx)? Or maybe it's not achievable?

you can pass your own jsx script file (or base64 encoded script instead), or you can even make a generator that will creatre a jsx script for your specific needs (with your parameteres)

also i think its possible to achieve it with expressions, (maybe at least by setting positions outside the frame, might work for some things)

**Comment by yarres** [comment]
Hi, thanks for your quick answer!
I will look into it.

**Comment by yarres** [comment]
Ok, so, I am now able to change the startTime of the layers I want to target via a jsx script, which is nice.

My next issue is, I would like to pass some data (the time at which the layers should start) from the nodejs api to the JSX script.

I have a small node.js server as in your docs that looks like that:
```

  const result = await client.addJob({    template: {      src: 'file:////Users/myMac/Documents/myComp.aep',      composition: 'TextInAndOut',      // frameStart: 120,      frameEnd:  300    },    assets: [      { type: 'script', src: "file:///Users/myName/CODE/updateComposition.jsx" },      {      src: `file:////Users/myName/Movies/Output/my_output_video.mp4`,      type: 'video',      layerName: "video.mp4"    },     {        type: "data",        layerName: "text1",        property: "Source Text",        value:"a new text"      },    {      type: "data",      layerName: "text2",      property: "Source Text",      value:"Another text"   },       ],    "actions": {  "postrender": [{        "module": "@nexrender/action-encode",        "preset": "mp4",        "output": "encoded_withAnim.mp4"      }, {        "module": "@nexrender/action-copy",        "input": "encoded_withAnim.mp4",        "output": `/Users/myName/Movies/Output/encoded_${Date.now()}.mp4`,      }]    }  });
```
Then in the `updateComposition.jsx` I do basically `composition.layer(index).startTime = myOffset;`
For now I hardcoded the value of `myOffset`  but I would like to get `myOffset` from the node api if possible.

**Comment by inlife** [comment]
well as I mentioned before, there are pretty much two ways with such an approach:
>you can pass your own jsx script file (or base64 encoded script instead), or you can even make a generator that will create a jsx script for your specific needs (with your parameters)

so first one would look smth like that:
```js

```
// your js side
let myOffset = 15;
let myScript = `composition.layer(index).startTime = ${myOffset};`
let encodedString = btoa(myScript);

```json
    {
        type: "script",
        src: `data:text/plain;base64,${encodedString}`
    },
```

The second one demands either creating a simple backend/micro service that will generate a script based on input params. Or usage of a custom pre-render action.

**Comment by yarres** [comment]
Many thanks! 

I wasn't aware of the possibility to pass the script as a base64.
I started to pass some data to the jsx file via prepending it to the jsx file which is not ideal. 
Your solution is better. I used a base64 encoded script with data passed via a js template string like you said.
